rust/src/derivation.rs

```diff
@@ -579,6 +540,177 @@ pub unsafe extern "C" fn zcashlc_get_sapling_receiver_for_unified_address(
+/// - The total size `seed_len` must be no larger than `isize::MAX`. See the safety documentation
+///   of `pointer::offset`.
+/// - Call [`zcashlc_free_account_metadata_key`] to free the memory associated with the returned
+///   pointer when done using it.
+#[no_mangle]
+pub unsafe extern "C" fn zcashlc_derive_account_metadata_key(
+    seed: *const u8,
+    seed_len: usize,
+    account: i32,
+    network_id: u32,
+) -> *mut ffi::AccountMetadataKey {
+    let res = catch_panic(|| {
+        let network = parse_network(network_id)?;
+        let seed = unsafe { slice::from_raw_parts(seed, seed_len) };
+        let account = zip32_account_index(account)?;
+
+        let key = zip32::registered::SecretKey::from_subpath(
+            b"MetadataKeys",
+            seed,
+            325,
+            &[
+                PathElement::new(ChildIndex::hardened(network.coin_type()), &[]),
+                PathElement::new(ChildIndex::hardened(account.into()), &[]),
+            ],
+        )?;
+
+        Ok(ffi::AccountMetadataKey::new(key))
+    });
+    unwrap_exc_or_null(res)
+}
+
+/// Derives a metadata key for private use from a ZIP 325 Account Metadata Key.
+///
+/// - `ufvk` is the external UFVK for which a metadata key is required, or `null` if the
+///   metadata key is "inherent" (for the same account as the Account Metadata Key).
+/// - `private_use_subject` is a globally unique non-empty sequence of at most 252 bytes
+///   that identifies the desired private-use context.
+///
+/// If `ufvk` is null, this function will return a single 32-byte metadata key.
+///
+/// If `ufvk` is non-null, this function will return one metadata key for every FVK item
+/// contained within the UFVK, in preference order. As UFVKs may in general change over
+/// time (due to the inclusion of new higher-preference FVK items, or removal of older
+/// deprecated FVK items), private usage of these keys should always follow preference
+/// order:
+/// - For encryption-like private usage, the first key in the array should always be
+///   used, and all other keys ignored.
+/// - For decryption-like private usage, each key in the array should be tried in turn
+///   until metadata can be recovered, and then the metadata should be re-encrypted
+///   under the first key.
+///
+/// # Safety
+///
+/// - `account_metadata_key` must be non-null and must point to a struct having the layout of
+///   [`ffi::AccountMetadataKey`].
+/// - If `ufvk` is non-null, it must point to a null-terminated UTF-8 string.
+/// - `private_use_subject` must be non-null and valid for reads for `private_use_subject_len`
+///   bytes.
+/// - The memory referenced by `private_use_subject` must not be mutated for the duration
+///   of the function call.
+/// - The total size `private_use_subject_len` must be no larger than `isize::MAX`. See
+///   the safety documentation of `pointer::offset`.
+/// - Call `zcashlc_free_boxed_slice` to free the memory associated with the returned
```

Comment:
```suggestion
/// - Call `zcashlc_free_txids` to free the memory associated with the returned
```

Also I still need to experiment to see if I can create aliases in Rust that are exposed as separate types in the C FFI (so I can rename `ffi::TxIds` for this new more general usage).